### PR TITLE
Fix registration for failed url

### PR DESCRIPTION
**Pull request by starkmsu**

### CODE REVIEW

**Review by dmitrysmall — COMMENTED** [review]
Unfortunately, can't add a comment to any lines outside the last modified block. I'll try to summorize several comments here.

1. Line 30: https://github.com/LykkeCity/MonitoringService/blob/8f3b65c0baacde3234d2c17afdd0a0a338ed5678/client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs#L30
"public static async" looks more conveniently.

2. Line 40:
https://github.com/LykkeCity/MonitoringService/blob/8f3b65c0baacde3234d2c17afdd0a0a338ed5678/client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs#L40
We return silently in case when auto-registration is off. May be, it's better to write some info to logs for explicitly understanding that there was no error, we just have such a settings.

3. Line 43: https://github.com/LykkeCity/MonitoringService/blob/8f3b65c0baacde3234d2c17afdd0a0a338ed5678/client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs#L43
Would it be better to use ArgumentNullException, like in other places.?

4. Line 55: https://github.com/LykkeCity/MonitoringService/blob/8f3b65c0baacde3234d2c17afdd0a0a338ed5678/client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs#L55
Looks like some kind of hardcode. The code will try to access "0.0.0.0" somehow if settings are not provided. Can it lead to crush? May be, there is a check up for the address inside of UrlMonitoringObjectModel constructor wich redirects "0.0.0.0" to somewhere else?

5. Line 70: https://github.com/LykkeCity/MonitoringService/blob/8f3b65c0baacde3234d2c17afdd0a0a338ed5678/client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs#L70
Returning silently if something goes wrong.

File: client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs [hunk 1]
```diff
@@ -69,11 +69,14 @@ public static class AutoRegistrationInMonitoring
                     if (monitoringRegistration.Url == myMonitoringUrl)
                         return;
 
-                    log.WriteMonitor("Auto-registration in monitoring", podTag, $"There is a registration for {myMonitoringName} in monitoring service!");
+                    if (monitoringRegistration.Url != _missingEnvVarUrl)
```
Comment (dmitrysmall) [inline client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs]: Seems like this If is guaranteed to be always true when we do not have a monitoring url provided in settings 'cause of the line 69. Also, this code is unreachable in the opposite case. Suppose, it's a desired flow, but should ask for sure :)
Btw, it's not so clear (for me personally) why does the `monitoringService.GetService(...)` return "0.0.0.0" as the url when monitoring has not been already registered. And does it mean that the constant `_missingEnvVarUrl` is defined twice in dofferent places?

Comment (dmitrysmall) [reply client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs]: Won't it be more clear with your intension if `GetService(...)` will return a tuple, where the first part will contain some `Enum` with monitoring registration (current) state, and the second part will be `monitoringRegistration` info?

File: client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs [hunk 2]
```diff
@@ -69,11 +69,14 @@ public static class AutoRegistrationInMonitoring
...
-                    myMonitoringUrl = _missingEnvVarUrl;
-                    string instanceTag = string.IsNullOrEmpty(podTag) ? Guid.NewGuid().ToString() : podTag;
-                    myMonitoringName = $"{myMonitoringName}-{instanceTag}";
+                        myMonitoringUrl = _missingEnvVarUrl;
+                        string instanceTag = string.IsNullOrEmpty(podTag) ? Guid.NewGuid().ToString() : podTag;
+                        myMonitoringName = $"{myMonitoringName}-{instanceTag}";
+                    }
                 }
                 catch (HttpOperationException)
                 {
```
Comment (dmitrysmall) [inline client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs:87]: Why do we need this catch (or why do we interpret it such a way) if duplicated registration is detected on the line 72?

**Review by dmitrysmall — COMMENTED** [review]
In general, the method's execution flow is not quite transparent for someone who looks at this code for the firts time. Or after ~6 months since he looked at it the last time :) For it's controlled by implicit conventions, constatcs comparations and logical structure optimization. May be, it would be usefull to add some comments in critical places.

File: client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs (hunk 2 above)
Comment (starkmsu) [reply client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs:87]: Service throws HttpOperationException with 404 when registration with such service name is not found

File: client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs (hunk 1 above)
Comment (starkmsu) [reply client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs]: "Seems like this If is guaranteed to be always true" - not true,
Line 69 prevents service from registration with the same data.
Line 72 handles the case when initiall there was no MyMonitoringUrl env var defined and this app was registered with _missingEnvVarUrl. So when this env var was added, proper url can replace  _missingEnvVarUrl. Without it not only original _missingEnvVarUrl will stay, additional registration with _missingEnvVarUrl will be added by Line 78

Comment (starkmsu) [reply client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs]: About suggestion of GetService result change - we can do it, but in another task.

**Review by dmitrysmall — APPROVED** [review]

File: client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs [hunk 3]
```diff
@@ -37,10 +36,13 @@ public static class AutoRegistrationInMonitoring
...
+            {
+                log.WriteMonitor("Auto-registration in monitoring", "", $"Auto-registration is disabled");
                 return;
+            }
 
             if (string.IsNullOrWhiteSpace(monitoringServiceUrl))
-                throw new ArgumentException("Argument is empty", nameof(monitoringServiceUrl));
+                throw new ArgumentNullException(nameof(monitoringServiceUrl));
 
             if (log == null)
```
Comment (dmitrysmall) [inline client/MonitoringServiceApiCaller/AutoRegistrationInMonitoring.cs]: Could you move it to the top of the method, please? For it is used above this line, at line 40.

**Review by dmitrysmall — APPROVED** [review]

**Pull request merged**